torchao/dtypes/int4_groupwise_preshuffle_tensor.py

```diff
@@ -0,0 +1,424 @@
+    packed_shape0, packed_shape1 = packed_weight.shape
+    new_shape = (packed_shape0, packed_shape1 * 2)
+    new = self.__class__(
+        packed_weight,
+        group_scale,
+        group_zero,
+        group_size=self.group_size,
+        shape_multiplier=self.shape_multiplier,
+        shape=new_shape,
+    )
+    return return_and_correct_aliasing(func, args, kwargs, new)
+
+
+@implements(aten.cat.default)
+def _(func, types, args, kwargs):
+    tensors, dim = fill_defaults(args, 2, [[], 0])
+    tensor_0 = tensors[0]
+    if dim < 0:
+        dim = dim + tensor_0.ndim
+
+    for i in range(1, len(tensors)):
+        assert tensor_0.packed_weight.ndim == tensors[i].packed_weight.ndim
+        assert tensor_0.group_scale.ndim == tensors[i].group_scale.ndim
+        assert tensor_0.group_zero.ndim == tensors[i].group_zero.ndim
+        assert tensor_0.group_size == tensors[i].group_size
+        assert tensor_0.shape_multiplier == tensors[i].shape_multiplier
+
+    packed_weight = [t.packed_weight for t in tensors]
+    group_scale = [t.group_scale for t in tensors]
+    group_zero = [t.group_zero for t in tensors]
+
+    # with group wise quantization, dimension of group_scale, packed_weight and
+    # origianl shape will be the same, so original dim argument applies
+    # to both packed_weight and group_scale
+    cat_packed_weight = aten.cat.default(packed_weight, dim)
+    if cat_packed_weight.ndim == 2:
+        sz_dim = 1 - dim
+    else:
+        sz_dim = dim
+
+    cat_group_scale = aten.cat.default(group_scale, sz_dim)
+    cat_group_zero = aten.cat.default(group_zero, sz_dim)
+    new_shape = list(cat_packed_weight.shape)
+    for i in range(len(tensor_0.shape_multiplier)):
+        new_shape[i] *= tensor_0.shape_multiplier[i]
+    new_shape = tuple(new_shape)
+    new = tensor_0.__class__(
+        cat_packed_weight,
+        cat_group_scale,
+        cat_group_zero,
+        group_size=tensor_0.group_size,
+        shape_multiplier=tensor_0.shape_multiplier,
+        shape=new_shape,
+    )
+    return return_and_correct_aliasing(func, args, kwargs, new)
+
+
+@implements(aten.transpose.int)
+def _(func, types, args, kwargs):
+    self, dim0, dim1 = args
+    packed_weight = self.packed_weight.transpose(dim0, dim1).contiguous()
+    shape_multiplier = self.shape_multiplier.copy()
+    shape_multiplier[dim0], shape_multiplier[dim1] = (
+        shape_multiplier[dim1],
+        shape_multiplier[dim0],
+    )
+
+    tensor_shape = list(packed_weight.shape)
+    for i in range(len(shape_multiplier)):
+        tensor_shape[i] *= shape_multiplier[i]
+    tensor_shape = tuple(tensor_shape)
+    new = self.__class__(
+        packed_weight,
+        self.group_scale,
+        self.group_zero,
+        self.group_size,
+        shape_multiplier,
+        tensor_shape,
+    )
+    return return_and_correct_aliasing(func, args, kwargs, new)
+
+
+to_int4_groupwise_preshuffle = Int4GroupwisePreshuffleTensor.from_float
```

Comment:
nit: why not just call the original function, to reduce the amount of indirection?

Reply:
OK, makes sense, this was also from previous design that we want to create some factory methods but it's not too important now I think. I can remove it